### PR TITLE
feat: Add the ability to specify "format as" buttons

### LINKED ISSUE #174: Add the ability to specify "format as" buttons
These would be documented under the _Customizing Download Links_ section of the docs.

Should be something like:

```
$this->Crud->action()->config('scaffold.index_formats', [
    'CSV' => Router::url(['_ext' => 'csv'])
]);
```

The above would simply add the `.csv` extension to the current page. It would be up to the user to define what renders for such a link.

The link should render in a new window/tab/whatever.

Refs #165

### PR DESCRIPTION
**Pull request by josegonzalez**
https://cl.ly/432j3J1g422b

Also includes a sample for integrating with CsvView.

Closes #174

### CODE REVIEW

File: docs/index-pages/customizing-the-index-page.rst [hunk 1]
```diff
@@ -280,8 +280,88 @@ Index Filters
...
+
+.. code-block:: php
+
+    public function index()
+    {
+        // only show the id, title, and created fields for csv output
+        if ($this->request->getParam('_ext') === 'csv') {
+            $this->set('_serialize', ['posts']);
+            $this->set('_extract', ['id', 'active', 'title', 'created']);
+        }
```
Comment (ADmad) [inline docs/index-pages/customizing-the-index-page.rst]: Should problem return here?

File: src/Template/Element/index/download_formats.ctp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?php
+if (empty($indexFormats)) {
+    return;
+}
+?>
+<div class="download-links">
+Download:
```
Comment (ADmad) [inline src/Template/Element/index/download_formats.ctp]: Wrap string in translation function perhaps.

File: docs/index-pages/customizing-the-index-page.rst (hunk 1 above)
Comment (josegonzalez) [inline docs/index-pages/customizing-the-index-page.rst]: No because its using Crud to implement the $posts variable.

**Comment by josegonzalez** [comment]
@ADmad rebased and fixed.

File: docs/index-pages/customizing-the-index-page.rst [hunk 3]
```diff
@@ -360,8 +360,97 @@ must be created in the ``PostsTable`` class.
...
+    use Cake\Routing\Router;
+
+    // link to the current page, except with extensions `json` or `xml`
+    // include the querystring argument as specified or you will lose any
+    // currently applied filters
+    $action = $this->Crud->action();
+    $action->config('scaffold.index_formats', [
+        [
+            'title' => 'JSON',
+            'url' => ['_ext' => 'json', '?' => $this->request->query]
```
Comment (ADmad) [inline docs/index-pages/customizing-the-index-page.rst]: The query property is deprecated, IIRC the new psr7 method is getQueryParams()

Comment (josegonzalez) [inline docs/index-pages/customizing-the-index-page.rst]: Should we require CakePHP 3.4+? That was something I was iffy on in docs/functionality, but would remove a ton of deprecated calls.

**Comment by josegonzalez** [comment]
Opened an issue for requiring CakePHP 3.4, #187.

**Pull request merged**